File: agent/ghidra_scripts/ExportAnalysis.java
//@category ReActAgent

import java.io.*;
import java.util.*;

import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.*;
import ghidra.program.model.listing.*;
import ghidra.program.model.symbol.*;
import ghidra.program.model.address.*;
import ghidra.program.model.mem.*;
import ghidra.program.model.data.*;

public class ExportAnalysis extends GhidraScript {
    private String outDir;

    @Override
    protected void run() throws Exception {
        String[] args = getScriptArgs();
        outDir = (args.length > 0) ? args[0] : "/tmp/ghidra_out";
        new File(outDir).mkdirs();
        Program prog = getCurrentProgram();
        if (prog == null) { println("[!] No program"); return; }
        println("[*] Exporting: " + prog.getName());
        exportFunctions(prog);
        exportDecompiled(prog);
        exportXrefs(prog);
        exportMemory(prog);
        exportCallgraph(prog);
        println("[*] Done -> " + outDir);
    }

    String esc(String s) {
        if (s == null) return "\"\"";
        StringBuilder sb = new StringBuilder("\"");
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == '"') sb.append("\\\"");
            else if (c == '\\') sb.append("\\\\");
            else if (c == '\n') sb.append("\\n");
            else if (c == '\r') sb.append("\\r");
            else if (c == '\t') sb.append("\\t");
            else sb.append(c);
        }
        sb.append("\"");
        return sb.toString();
    }

    void save(String fn, String data) throws Exception {
        try (PrintWriter w = new PrintWriter(new File(outDir, fn))) { w.print(data); }
    }

    void exportFunctions(Program prog) throws Exception {
        FunctionManager fm = prog.getFunctionManager();
        StringBuilder sb = new StringBuilder("[\n");
        boolean first = true;
        for (Function f : fm.getFunctions(true)) {
            if (!first) sb.append(",\n"); first = false;
            sb.append("  {\"name\":").append(esc(f.getName()));
            sb.append(",\"address\":").append(esc(f.getEntryPoint().toString()));
            sb.append(",\"size\":").append(f.getBody().getNumAddresses());
            sb.append(",\"signature\":").append(esc(f.getSignature().toString()));
            String cc = f.getCallingConventionName();
            sb.append(",\"cconv\":").append(esc(cc != null ? cc : ""));
            sb.append(",\"no_return\":").append(f.hasNoReturn());
            sb.append(",\"external\":").append(f.isExternal());
            String cmt = f.getComment();
            sb.append(",\"comment\":").append(esc(cmt != null ? cmt : ""));
            sb.append("}");
        }
        sb.append("\n]");
        save("ghidra_functions.json", sb.toString());
        println("[+] functions: " + fm.getFunctionCount());
    }

    void exportDecompiled(Program prog) throws Exception {
        FunctionManager fm = prog.getFunctionManager();
        DecompInterface dec = new DecompInterface();
        dec.openProgram(prog);
        StringBuilder sb = new StringBuilder("{\n");
        boolean first = true;
        for (Function f : fm.getFunctions(true)) {
            if (f.isExternal()) continue;
            if (!first) sb.append(",\n"); first = false;
            String addr = f.getEntryPoint().toString();
            sb.append("  ").append(esc(addr)).append(":{\"addr\":").append(esc(addr));
            sb.append(",\"name\":").append(esc(f.getName()));
            DecompileResults r = dec.decompileFunction(f, 60, monitor);
            String code = (r != null && r.getDecompiledFunction() != null)
                ? r.getDecompiledFunction().getC() : "(failed)";
            sb.append(",\"code\":").append(esc(code));
            sb.append("}");
        }
        sb.append("\n}");
        save("ghidra_decompiled.json", sb.toString());
        println("[+] decompiled done");
        dec.dispose();
    }

    void exportXrefs(Program prog) throws Exception {
        FunctionManager fm = prog.getFunctionManager();
        ReferenceManager rm = prog.getReferenceManager();
        StringBuilder sb = new StringBuilder("{\n");
        boolean first = true;
        for (Function f : fm.getFunctions(true)) {
            if (f.isExternal()) continue;
            String addr = f.getEntryPoint().toString();
            ReferenceIterator it = rm.getReferencesTo(f.getEntryPoint());
            if (!it.hasNext()) continue;
            if (!first) sb.append(",\n"); first = false;
            sb.append("  ").append(esc(addr)).append(":[\n");
            boolean f2 = true;
            while (it.hasNext()) {
                Reference ref = it.next();
                if (!f2) sb.append(",\n"); f2 = false;
                sb.append("    {\"from\":").append(esc(ref.getFromAddress().toString()));
                sb.append(",\"type\":").append(esc(ref.getReferenceType().toString()));
                sb.append("}");
            }
            sb.append("\n  ]");
        }
        sb.append("\n}");
        save("ghidra_xrefs.json", sb.toString());
        println("[+] xrefs done");
    }

    void exportMemory(Program prog) throws Exception {
        Memory mem = prog.getMemory();
        StringBuilder sb = new StringBuilder("[\n");
        boolean first = true;
        for (MemoryBlock b : mem.getBlocks()) {
            if (!first) sb.append(",\n"); first = false;
            sb.append("  {\"name\":").append(esc(b.getName()));
            sb.append(",\"start\":").append(esc(b.getStart().toString()));
            sb.append(",\"end\":").append(esc(b.getEnd().toString()));
            sb.append(",\"size\":").append(b.getSize());
            String p = (b.isRead()?"r":"")+(b.isWrite()?"w":"")+(b.isExecute()?"x":"");
            sb.append(",\"perms\":").append(esc(p));
            sb.append(",\"init\":").append(b.isInitialized());
            sb.append("}");
        }
        sb.append("\n]");
        save("ghidra_memory.json", sb.toString());
        println("[+] memory: " + mem.getBlocks().length + " blocks");
    }

    void exportCallgraph(Program prog) throws Exception {
        FunctionManager fm = prog.getFunctionManager();
        Listing listing = prog.getListing();
        StringBuilder sb = new StringBuilder("{\n");
        boolean first = true;
        for (Function f : fm.getFunctions(true)) {
            if (f.isExternal()) continue;
            String caller = f.getEntryPoint().toString();
            java.util.List<String> c = new java.util.ArrayList<>();
            AddressSetView body = f.getBody();
            AddressIterator ai = body.getAddresses(true);
            while (ai.hasNext()) {
                Address a = ai.next();
                CodeUnit cu = listing.getCodeUnitAt(a);
                if (cu == null) continue;
                for (Reference ref : cu.getReferencesFrom()) {
                    String t = ref.getReferenceType().toString();
                    if (t.contains("CALL") || t.contains("JUMP")) {
                        Address to = ref.getToAddress();
                        Function tf = fm.getFunctionAt(to);
                        c.add("{\"to\":" + esc(to.toString()) +
                            ",\"name\":" + esc(tf != null ? tf.getName() : "unknown") +
                            ",\"type\":" + esc(t) + "}");
                    }
                }
            }
            if (c.isEmpty()) continue;
            if (!first) sb.append(",\n"); first = false;
            sb.append("  ").append(esc(caller)).append(":[\n");
            for (int i = 0; i < c.size(); i++) {
                if (i > 0) sb.append(",\n");
                sb.append("    ").append(c.get(i));
            }
            sb.append("\n  ]");
        }
        sb.append("\n}");
        save("ghidra_callgraph.json", sb.toString());
        println("[+] callgraph done");
    }
}

File: agent/ghidra_scripts/export_analysis.py
#@category ReActAgent
#@menupath ReActAgent.ExportAnalysis
"""
Ghidra headless script: Export binary analysis data as JSON.
Run with: analyzeHeadless <project_dir> <project_name> -import <binary> -scriptPath <dir> -postScript export_analysis.py <output_dir>

Outputs JSON files used by the ReAct Agent's Ghidra tools.
"""

import json
import os
import sys

from ghidra.program.model.listing import Function, FunctionManager
from ghidra.program.model.symbol import SourceType
from ghidra.program.model.address import Address
from ghidra.util.task import ConsoleTaskMonitor

# Get output directory from script arguments
output_dir = sys.argv[1] if len(sys.argv) > 1 else os.getcwd()
os.makedirs(output_dir, exist_ok=True)

monitor = ConsoleTaskMonitor()
program = getCurrentProgram()
if program is None:
    print("[!] No program loaded")
    sys.exit(1)

listing = program.getListing()
function_manager = program.getFunctionManager()
addr_factory = program.getAddressFactory()

name = program.getName()
print("[*] Exporting analysis for: {}".format(name))


# ---- Functions ----
def export_functions():
    functions = []
    func = function_manager.getFunctions(True)
    for f in func:
        entry = {
            "name": f.getName(),
            "address": str(f.getEntryPoint()),
            "size": f.getBody().getNumAddresses(),
            "signature": str(f.getSignature()),
            "calling_convention": str(f.getCallingConventionName()),
            "has_no_return": f.hasNoReturn(),
            "is_external": f.isExternal(),
            "comment": f.getComment(),
        }
        functions.append(entry)

    with open(os.path.join(output_dir, "ghidra_functions.json"), "w") as fp:
        json.dump(functions, fp, indent=2)
    print("[+] Exported {} functions".format(len(functions)))
    return functions


# ---- Decompilation ----
def decompile_function(func_addr_str):
    """Decompile a specific function by address."""
    try:
        addr = addr_factory.getAddress(func_addr_str)
        func = function_manager.getFunctionAt(addr)
        if func is None:
            return {"error": "Function not found at " + func_addr_str}

        from ghidra.app.decompiler import DecompInterface
        decomp = DecompInterface()
        decomp.openProgram(program)

        results = decomp.decompileFunction(func, 60, monitor)
        if results and results.getDecompiledFunction():
            code = results.getDecompiledFunction().getC()
            return {"address": func_addr_str, "name": func.getName(), "decompiled": code}
        else:
            return {"address": func_addr_str, "name": func.getName(), "decompiled": "(decompilation failed)"}
    except Exception as e:
        return {"error": str(e), "address": func_addr_str}


def export_all_decompilations():
    decomp_results = {}
    func = function_manager.getFunctions(True)
    for f in func:
        if f.isExternal():
            continue
        addr = str(f.getEntryPoint())
        decomp_results[addr] = decompile_function(addr)

    with open(os.path.join(output_dir, "ghidra_decompiled.json"), "w") as fp:
        json.dump(decomp_results, fp, indent=2)
    print("[+] Exported {} decompilations".format(len(decomp_results)))
    return decomp_results


# ---- Cross References ----
def export_xrefs():
    xrefs = {}
    func = function_manager.getFunctions(True)
    for f in func:
        if f.isExternal():
            continue
        addr_str = str(f.getEntryPoint())
        entry_xrefs = []

        # XREFS TO this function
        refs_to = program.getListing().getCodeUnits(f.getBody(), True)
        # Actually use reference manager
        ref_mgr = program.getReferenceManager()
        addr = f.getEntryPoint()
        refs = ref_mgr.getReferencesTo(addr)
        for ref in refs:
            from_addr = ref.getFromAddress()
            ref_type = str(ref.getReferenceType())
            entry_xrefs.append({
                "type": "to",
                "from": str(from_addr),
                "ref_type": ref_type
            })

        if entry_xrefs:
            xrefs[addr_str] = entry_xrefs

    with open(os.path.join(output_dir, "ghidra_xrefs.json"), "w") as fp:
        json.dump(xrefs, fp, indent=2)
    print("[+] Exported xrefs for {} functions".format(len(xrefs)))
    return xrefs


# ---- Data Types ----
def export_data_types():
    dt_mgr = program.getDataTypeManager()
    types = []
    for dt in dt_mgr.getAllDataTypes():
        try:
            types.append({
                "name": str(dt.getName()),
                "size": dt.getLength(),
                "category": str(dt.getCategoryPath()),
                "type": str(type(dt).__name__),
            })
        except:
            pass

    with open(os.path.join(output_dir, "ghidra_datatypes.json"), "w") as fp:
        json.dump(types, fp, indent=2)
    print("[+] Exported {} data types".format(len(types)))


# ---- Memory Map ----
def export_memory():
    mem_blocks = []
    mem = program.getMemory()
    for block in mem.getBlocks():
        entry = {
            "name": block.getName(),
            "start": str(block.getStart()),
            "end": str(block.getEnd()),
            "size": block.getSize(),
            "permissions": "{}{}{}".format(
                "r" if block.isRead() else "",
                "w" if block.isWrite() else "",
                "x" if block.isExecute() else "",
            ),
            "initialized": block.isInitialized(),
        }
        mem_blocks.append(entry)

    with open(os.path.join(output_dir, "ghidra_memory.json"), "w") as fp:
        json.dump(mem_blocks, fp, indent=2)
    print("[+] Exported {} memory blocks".format(len(mem_blocks)))


# ---- Call Graph ----
def export_call_graph():
    calls = {}
    func = function_manager.getFunctions(True)
    for f in func:
        if f.isExternal():
            continue
        caller_addr = str(f.getEntryPoint())
        callees = []
        body = f.getBody()
        code_units = listing.getCodeUnits(body, True)
        for cu in code_units:
            for ref in cu.getReferencesFrom():
                to_addr = ref.getToAddress()
                ref_type = str(ref.getReferenceType())
                if "CALL" in ref_type.upper() or "JUMP" in ref_type.upper():
                    to_func = function_manager.getFunctionAt(to_addr)
                    callees.append({
                        "to": str(to_addr),
                        "to_name": to_func.getName() if to_func else "unknown",
                        "type": ref_type,
                    })
        if callees:
            calls[caller_addr] = callees

    with open(os.path.join(output_dir, "ghidra_callgraph.json"), "w") as fp:
        json.dump(calls, fp, indent=2)
    print("[+] Exported call graph: {} callers".format(len(calls)))


# ---- Main ----
export_functions()
export_all_decompilations()
export_xrefs()
export_data_types()
export_memory()
export_call_graph()

print("[*] Ghidra analysis export complete -> {}".format(output_dir))
